### PR TITLE
MacroSubstitution Stmt and Decl covering methods

### PR DESCRIPTION
**Pull request by PappasBrent**
Add methods to MacroSubstitution to get the lowest Stmt and Decl that the substitution completely covers.
Change equality comparisons between Macro objects to use the RawMacro() method

### CODE REVIEW

File: lib/AST/Macro.cpp [hunk 1]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+std::pair<std::optional<Token>, std::optional<Token>>
+GetFirstAndLastFullySubstitutedTokens(
+  MacroSubstitution const &substitution) noexcept {
+
+  // Get the fully substituted first token
+  std::optional<Token> first_tok = std::nullopt;
+  {
+    auto subst = std::optional(substitution);
+    auto first_replacement_child = std::optional<Macro>();
+    while (subst && !subst->ReplacementChildren().empty()) {
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:894]: Save return value of `ReplacementChildren()` to a local var so you can reuse it.

File: lib/AST/Macro.cpp [hunk 2]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
 
+// Returns the first and last fully substituted tokens from a given macro
+// substitution, if they exist.
+std::pair<std::optional<Token>, std::optional<Token>>
+GetFirstAndLastFullySubstitutedTokens(
+  MacroSubstitution const &substitution) noexcept {
+
+  // Get the fully substituted first token
+  std::optional<Token> first_tok = std::nullopt;
+  {
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:891]: What is this for? Ensuring destruction of something inside of the `{`?

File: lib/AST/Macro.cpp [hunk 3]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+    auto subst = std::optional(substitution);
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:892]:
```c++
std::optional<MacroSubstitution> subst = substitution;
```

File: lib/AST/Macro.cpp [hunk 4]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+      if (first_macro_tok) {
+        first_tok = std::optional(first_macro_tok->ParsedLocation());
+      }
+    }
+  }
+
+  // Get the last substituted first token
+  std::optional<Token> last_tok = std::nullopt;
+  {
+    auto subst = std::optional(substitution);
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:910]: Reuse `subst` from before?

File: lib/AST/Macro.cpp [hunk 5]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+    auto last_replacement_child = std::optional<Macro>();
+    while (subst && !subst->ReplacementChildren().empty()) {
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:912]: Refactor to share the same `MacroRange` returned from `subst->ReplacementChildren()`.

File: lib/AST/Macro.cpp [hunk 6]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+      // Traverse the next end token
+      last_replacement_child = *(subst->ReplacementChildren().end() - 1);
+      subst = MacroSubstitution::From(*last_replacement_child);
+    }
+    if (last_replacement_child) {
+      const auto last_macro_tok = MacroToken::From(*last_replacement_child);
+      if (last_macro_tok) {
+        last_tok = std::optional(last_macro_tok->ParsedLocation());
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:920]: Don't need `std::optional` here.

File: lib/AST/Macro.cpp [hunk 7]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+    auto first_replacement_child = std::optional<Macro>();
+    while (subst && !subst->ReplacementChildren().empty()) {
+      // Traverse the next begin token
+      first_replacement_child = *subst->ReplacementChildren().begin();
+      subst = MacroSubstitution::From(*first_replacement_child);
+    }
+    if (first_replacement_child) {
+      const auto first_macro_tok = MacroToken::From(*first_replacement_child);
+      if (first_macro_tok) {
+        first_tok = std::optional(first_macro_tok->ParsedLocation());
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:902]: Don't need `std::optional` here.

File: lib/AST/Macro.cpp [hunk 8]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+    GetFirstAndLastFullySubstitutedTokens(*this);
+
+  // Check that the substitution actually expands to at least one token.
+  if (!(first_tok && last_tok)) {
+    return std::nullopt;
+  }
+
+  // Walk up the tokens' context tree until we find a Stmt whose first and
+  // last tokens are the first and last fully expanded tokens.
+  auto ctx = first_tok->Context();
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:943]: Convert to `for` loop.

File: lib/AST/Macro.cpp [hunk 9]
```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+    GetFirstAndLastFullySubstitutedTokens(*this);
+
+  // Check that the substitution actually expands to at least one token.
+  if (!(first_tok && last_tok)) {
+    return std::nullopt;
+  }
+
+  // Walk up the tokens' context tree until we find a Decl whose first and
+  // last tokens are the first and last fully expanded tokens.
+  auto ctx = first_tok->Context();
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:971]: Convert to for loop.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: lib/AST/Macro.cpp (hunk 2 above)
Comment (PappasBrent) [reply lib/AST/Macro.cpp:891]: Honestly it was to make it easier for me to copy/paste code between the two blocks.

Comment (PappasBrent) [reply lib/AST/Macro.cpp:891]: The block are ultimately unnecessary if I reuse existing declarations so I'll remove them

File: lib/AST/Macro.cpp (hunk 6 above)
Comment (PappasBrent) [reply lib/AST/Macro.cpp:920]: Why not? If a macro expands to no tokens then is it possible for the the first/last token to be nothing?

File: lib/AST/Macro.cpp [hunk 10]
```diff
@@ -880,6 +881,165 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+GetFirstAndLastFullySubstitutedTokens(MacroSubstitution const &sub) noexcept {
+  auto subst_replacement_children = sub.ReplacementChildren();
+
+  // Higher-order function for getting the left/right corner of the substitution
+  // tree.
+  // The walk parameter tells us how to traverse the current substitution's
+  // replacement children of tree to go left/right.
+  // Returns an optional Token because the top-level substitution may not expand
+  // to any tokens.
+  const auto Corner = [&subst_replacement_children]
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:896]: Extract into `static` function template, parameterized on corner function/functor type.

File: lib/AST/Macro.cpp [hunk 11]
```diff
@@ -880,6 +881,165 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+  const auto Right = [](MacroRange range) { return *(range.end() - 1); };
+  const auto first_tok = Corner(Left);
+  const auto last_tok = Corner(Right);
+
+  return std::pair(first_tok, last_tok);
+}
+
+// Returns the index of the next final expansion or file token in the given
+// token list after the given token
+std::optional<uint64_t>
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:932]: `static`

File: lib/AST/Macro.cpp [hunk 12]
```diff
@@ -880,6 +881,165 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
   }
 }
 
+// Returns the first and last fully substituted tokens from a given macro
+// substitution, if they exist.
+std::pair<std::optional<Token>, std::optional<Token>>
+GetFirstAndLastFullySubstitutedTokens(MacroSubstitution const &sub) noexcept {
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:887]: `static`, remove `Get` prefix.

File: lib/AST/Macro.cpp [hunk 13]
```diff
@@ -880,6 +881,165 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+  return std::pair(first_tok, last_tok);
+}
+
+// Returns the index of the next final expansion or file token in the given
+// token list after the given token
+std::optional<uint64_t>
+NextFinalExpansionOrFileTokenIndex(const std::vector<pasta::TokenImpl> &tokens,
+                                   const Token &tok) {
+  for (uint64_t i = tok.Index() + 1; i < tokens.size(); i++)
+  {
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:936]: `{` on previous line.

File: lib/AST/Macro.cpp [hunk 14]
```diff
@@ -880,6 +881,165 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
...
+    const auto next_tok = tokens.at(i);
+    const auto next_tok_role = next_tok.Role();
+    if (next_tok_role == TokenRole::kFinalMacroExpansionToken ||
+        next_tok_role == TokenRole::kFileToken)
+    {
```
Comment (pgoodman) [inline lib/AST/Macro.cpp:941]: `{` on previous line.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: lib/AST/Macro.cpp (hunk 10 above)
Comment (PappasBrent) [reply lib/AST/Macro.cpp:896]: Ah cool

File: lib/AST/Macro.cpp (hunk 13 above)
Comment (PappasBrent) [reply lib/AST/Macro.cpp:936]: oops

**Pull request merged**